mapper/pojo-base/src/main/java/org/hibernate/search/mapper/pojo/massindexing/impl/PojoMassIndexingBatchCoordinator.java

```diff
@@ -188,10 +188,13 @@ private void afterBatch() throws InterruptedException {
 			);
 		}
 		flushAndRefresh();
-		applyToAllContexts(
-				context -> context.agent().preStop()
-						.thenRun( () -> context.agent().stop() )
-		);
+		// Not using the `applyToAllContexts(..)` here as we want to make sure that stop() is executed right after preStop()
+		// completes and before going to another agent.
+		// Otherwise, it might cause unexpected failures when running with certain DBs (Oracle), see HSEARCH-4773
```

Comment:
Is the problem related to locks, i.e. we're concurrently changing the same table from multiple connections and Oracle returns a "concurrent change" error?

If so, maybe we should rather try to avoid problematic "read then update" patterns such as this one:

https://github.com/hibernate/hibernate-search/blob/75d0dcdbeff51067c600837202b2b01a3de9d228/mapper/orm-coordination-outbox-polling/src/main/java/org/hibernate/search/mapper/orm/coordination/outboxpolling/cluster/impl/AgentPersister.java#L55-L65


A simple delete-by-ID (without a "find") might do the trick here, and maybe it would solve your problem?

I'm suggesting this because I suspect we could have the same problem in other situations (e.g. when shutting down Hibernate Search and removing outbox event processor agents)